rt: clarify the documentation of `Runtime::spawn`

Its-Just-Nans opened the pull request:
## Motivation

Following https://github.com/tokio-rs/tokio/pull/7802#pullrequestreview-3614952466

I think clarifiying the doc of `spawn()` is better

[Rendered](https://deploy-preview-7803--tokio-rs.netlify.app/tokio/)

Its-Just-Nans commented on tokio/src/runtime/runtime.rs:
```diff
@@ -200,15 +200,14 @@ impl Runtime {
     }
 
-    /// Spawns a future onto the Tokio runtime.
+    /// Spawns a future onto the Tokio runtime and return a [`JoinHandle`].
```
Yes, it was for consistency
https://github.com/tokio-rs/tokio/blob/0a3e3862697d902cffd729506cb28f6007c204d1/tokio/src/task/spawn.rs#L8-L9

Its-Just-Nans commented:
> By the way, there are `Handle::spawn` and `tokio::spawn`, could you also update their docs?

I've updated, tell me what I need to change

ADD-SP reviewed (commented):
We don't need fully qualified syntax in the docs.

<img src="https://github.com/user-attachments/assets/a6be48c8-e5e1-41ed-982b-9408f0798d62" />

<img src="https://github.com/user-attachments/assets/7df0aed8-3e23-44cd-809b-206e44efa166" />

Its-Just-Nans commented on tokio/src/runtime/handle.rs:
```diff
@@ -164,9 +164,8 @@ impl Handle {
-    /// when `spawn` is called, even if you don't await the returned
-    /// `JoinHandle`.
+    /// Executes the given future. Depending on the current [`crate::runtime::Runtime`],
+    /// it may start immediately in the background or require blocking on the [`JoinHandle`].
```
I'm not sure to understand the changes you want @ADD-SP  😅

martin-g replied:
IMO it will be clearer if the documentation explain when it will be executed immediately and when it needs blocking. I.e. be more specific instead of letting the reader guess and experiment.

ADD-SP replied:
> IMO it will be clearer if the documentation explain when it will be executed immediately and when it needs blocking. I.e. be more specific instead of letting the reader guess and experiment.

However, this would make the documentation increasingly lengthy, and no one enjoys reading academic papers within `docs.rs`. That's why [I recommend keeping the original documentation as it stands](https://github.com/tokio-rs/tokio/pull/7802#pullrequestreview-3614952466)—the confusion it has caused so far remains within acceptable range.

Its-Just-Nans replied:
I've updated my branch

Thanks

martin-g commented on tokio/src/task/spawn.rs:
```diff
@@ -6,16 +6,17 @@ use std::future::Future;
     /// The provided future will start running in the background immediately
     /// when `spawn` is called, even if you don't await the returned
-    /// `JoinHandle`.
+    /// [`JoinHandle`](crate::task::JoinHandle).
```
```suggestion
    /// [JoinHandle].
```

martin-g commented on tokio/src/task/spawn.rs:
```diff
@@ -6,16 +6,17 @@ use std::future::Future;
     /// different thread to be executed. The specifics depend on the current
-    /// [`Runtime`](crate::runtime::Runtime) configuration.
+    /// [`Runtime`](crate::runtime::Runtime) configuration - it may start immediately in
+    /// the background or require blocking on the [`JoinHandle`](crate::task::JoinHandle).
```
```suggestion
    /// the background or require blocking on the [JoinHandle].
```

martin-g commented on tokio/src/runtime/runtime.rs:
```diff
@@ -206,9 +206,8 @@ impl Runtime {
-    /// The provided future will start running in the background immediately
-    /// when `spawn` is called, even if you don't await the returned
-    /// `JoinHandle`.
+    /// Executes the given future. Depending on the current [`Runtime`] configuration,
```
```suggestion
    /// Executes the given future. Depending on the current [Runtime] configuration,
```

martin-g commented on tokio/src/runtime/handle.rs:
```diff
@@ -164,9 +164,8 @@ impl Handle {
-    /// The provided future will start running in the background immediately
-    /// when `spawn` is called, even if you don't await the returned
-    /// `JoinHandle`.
+    /// Executes the given future. Depending on the current [`Runtime`](crate::runtime::Runtime),
```
```suggestion
    /// Executes the given future. Depending on the current [Runtime],
```

Its-Just-Nans replied:
Thanks - fixed, I'm going with the markdown link

martin-g commented on tokio/src/runtime/handle.rs:
```diff
@@ -164,9 +164,8 @@ impl Handle {
-    /// when `spawn` is called, even if you don't await the returned
-    /// `JoinHandle`.
+    /// Executes the given future. Depending on the current [`Runtime`](crate::runtime::Runtime),
+    /// it may start immediately in the background or require blocking on the [`JoinHandle`].
```
```suggestion
    /// it may start immediately in the background or require blocking on the [JoinHandle].
```

martin-g reviewed (commented):
I've updated the infra-doc links according to https://doc.rust-lang.org/rustdoc/write-documentation/linking-to-items-by-name.html
> Backticks around the link will be stripped, so [`Option`] will correctly link to Option.

Its-Just-Nans commented on tokio/src/task/spawn.rs:
```diff
@@ -6,16 +6,17 @@ use std::future::Future;
 cfg_rt! {
     /// Spawns a new asynchronous task, returning a
-    /// [`JoinHandle`](JoinHandle) for it.
+    /// [`JoinHandle`](crate::task::JoinHandle) for it.
```
But now the spellcheck is annoying :/

Darksonn commented on tokio/src/runtime/runtime.rs:
```diff
@@ -206,9 +206,8 @@ impl Runtime {
-    /// when `spawn` is called, even if you don't await the returned
-    /// `JoinHandle`.
+    /// Executes the given future. Depending on the current [Runtime] configuration,
+    /// it may start immediately in the background or require blocking on the [JoinHandle].
```
This is wrong!

It never requires blocking on the `JoinHandle`.

As long as `Runtime::block_on()` is called *with any future*, spawned tasks will execute. The `JoinHandle` does not matter. You could even call `rt.block_on(pending())` in a background thread.

Its-Just-Nans replied:
Yes
thanks

Its-Just-Nans commented:
> And then the word "running" can be a link to a new section in [the docs for `tokio::runtime`](https://docs.rs/tokio/latest/tokio/runtime/index.html), which clarifies that the multi-threaded runtime is always running, that the current-thread runtime is only running when there is at least on call to `Runtime::block_on()`, and that a `LocalSet` is running when it's being driven by one of more mechanisms.

I think that's a nice idea! I've updated my branch with this change(s)


ping @ADD-SP - i've updated the docs

Darksonn commented on tokio/src/runtime/handle.rs:
```diff
@@ -164,13 +164,13 @@ impl Handle {
-    /// when `spawn` is called, even if you don't await the returned
-    /// `JoinHandle`.
+    /// Executes the given future. Depending if the current [`Runtime`](runtime::Runtime)
+    /// is [running][running-runtime] it may start immediately in the background.
```
I don't like this wording. This paragraph was added to make it clear that awaiting the `JoinHandle` is *not* required, but you removed that.

Darksonn commented on tokio/src/runtime/mod.rs:
```diff
@@ -205,6 +205,17 @@
 //! This is done with [`Builder::enable_io`] and [`Builder::enable_time`]. As a
 //! shorthand, [`Builder::enable_all`] enables both resource drivers.
 //!
+//! #### Running runtime
```
You placed this under "Runtime Configurations", but it's not a way you can configure the runtime. Please make it into its own section.

Also, a different section name such as "Driving the runtime" would make more sense to me.

Its-Just-Nans replied:
Thanks I edited the docs

Darksonn commented on tokio/src/runtime/mod.rs:
```diff
@@ -205,6 +205,17 @@
+//! A multi-threaded runtime is always running. The current-thread runtime is only
+//! running when there is at least on call to [`Runtime::block_on`]. A
+//! [`LocalSet`](crate::task::LocalSet) is running when it's being driven by one
+//! of more mechanisms.
```
I think this wording is confusing. Perhaps something along the lines of

> A Tokio runtime can only execute tasks if the runtime is running. Normally this is not an issue as the default configuration of a runtime is always running, but alternate configurations such as the current-thread runtime require that `Runtime::block_on` is called.
>
> * A multi-threaded runtime is always running because it spawns its own worker threads.
> * A current-thread runtime does not spawn any worker threads, so it can only execute tasks when you provide a thread by calling [`Runtime::block_on`].
> * A [`LocalSet`] only executes local tasks spawned on it when the `LocalSet` is `.await`ed or otherwise driven using one of its methods for this purpose.
> 
> Please be aware that [`Handle::block_on`] does *not* drive the runtime. There must be at least one call to [`Runtime::block_on`] when using the current-thread runtime. [`Handle::block_on`] is not enough.

Its-Just-Nans commented:
Hi

Should I change something ?

Darksonn commented on tokio/src/task/spawn.rs:
```diff
@@ -6,16 +6,20 @@ use std::future::Future;
+    /// [running runtime][running-runtime], it may start immediately in the background,
+    /// even if you don't await the returned [`JoinHandle`].
+    /// On a blocked runtime, the user must drive the runtime forward (for example,
+    /// by calling [`Runtime::block_on`](crate::runtime::Runtime::block_on)).
```
This wording is also confusing. You say "in a running runtime, it may start", which implies that even if it's running, it *might not* start. Please avoid the word "may".

Its-Just-Nans commented on tokio/src/task/spawn.rs:
```diff
@@ -6,16 +6,20 @@ use std::future::Future;
-    /// [`Runtime`](crate::runtime::Runtime) configuration.
+    /// [`Runtime`](crate::runtime::Runtime) configuration. In a
+    /// [running runtime][running-runtime], the task will start immediately in the
+    /// background, even if you don't await the returned [`JoinHandle`].
```
Okay I removed the await part

Darksonn reviewed (approved):
Thank you.
